pkg/miniogw/multipart.go

```diff
@@ -0,0 +1,438 @@
+package miniogw
+
+import (
+	"context"
+	"io"
+	"sort"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	minio "github.com/minio/minio/cmd"
+	"github.com/minio/minio/pkg/hash"
+	"storj.io/storj/pkg/paths"
+	"storj.io/storj/pkg/storage/objects"
+)
+
+func (s *storjObjects) NewMultipartUpload(ctx context.Context, bucket, object string, metadata map[string]string) (uploadID string, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	uploads := s.storj.multipart
+
+	upload, err := uploads.Create(bucket, object, metadata)
+	if err != nil {
+		return "", err
+	}
+
+	objectStore, err := s.storj.bs.GetObjectStore(ctx, bucket)
+	if err != nil {
+		uploads.mu.Lock()
+		delete(uploads.pending, upload.ID)
+		uploads.mu.Unlock()
```

Comment:
Should we encapsulate all operations that require acquiring the lock into the MultipartUploads type? It looks error prone to me to deal with the lock outside of that type.

Reply:
done